MobiusCore/Source/EffectHandlers/EffectHandlingConnection.swift

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+class EffectHandlingConnection<Effect, Event>: Disposable {
+    private let handleEffect: (Effect, Response<Event>) -> Disposable?
+
+    private let lock = Lock()
+    // Keep track of each received effect's state.
+    // When an effect has completed, it should be removed from this dictionary.
+    // When disposing this effect handler, all entries must be removed.
+    private var handlingEffects: [UUID: EffectHandlingState<Event>] = [:]
+    private var unsafeOutput: Consumer<Event>?
+
+    init(
+        handleInput: @escaping (Effect, Response<Event>) -> Disposable?,
+        output unsafeOutput: @escaping Consumer<Event>
+    ) {
+        self.handleEffect = handleInput
+        self.unsafeOutput = unsafeOutput
+    }
+
+    func handle(_ effect: Effect) -> Bool {
+        let id = UUID()
+        let response = Response(onSend: self.output, onEnd: { [unowned self] in self.delete(id: id) })
+        let effectHandlerState: EffectHandlingState<Event> = .unhandled(response: response)
+
+        create(id: id, state: effectHandlerState)
```

Comment:
Calling `create` here is costly and technically unnecessary.
Ideally we should only store the state once handleEffect(effect, response) has returned. The problem though, is that `handleEffect` could call `response.end` before it returns. In that case, it should not be stored.

Any other ideas here would be appreciated.

Reply:
Sounds like you need to smuggle state on a side channel. :-)

One option would be to make `id` a `var` and clear it in `onEnd`. `create` would then only be called if it’s non-nil after `handleEffect`.